src/Compiler/Microsoft.CodeAnalysis.Razor/src/ComponentDetectionConventions.cs

```diff
@@ -28,21 +25,22 @@ public static bool IsComponent(INamedTypeSymbol symbol, INamedTypeSymbol icompon
             symbol.AllInterfaces.Contains(icomponentSymbol);
     }
 
-    public static bool IsComponent(INamedTypeSymbol symbol, string icomponentSymbolName)
+    public static bool IsComponent(INamedTypeSymbol typeSymbol, string icomponentSymbolName)
     {
-        if (symbol is null)
+        if (typeSymbol.DeclaredAccessibility != Accessibility.Public ||
+            typeSymbol.IsAbstract)
         {
-            throw new ArgumentNullException(nameof(symbol));
+            return false;
         }
 
-        if (icomponentSymbolName is null)
+        foreach (var interfaceSymbol in typeSymbol.AllInterfaces)
```

Comment:
The `AllInterfaces` member allocates on every call. The internal array is cached but our `GetPublicSymbols` call allocates every time. I looked through Razor and it appears we do this check quite a bit (just do a quick search for `AllInterfaces` in the code). Occurs to me this might be a place to consider for future performance changes. Essentially plumb a better API into Roslyn and use it in Razor. 

Believe this is the best we can do with the current API surface area.

Reply:
> Believe this is the best we can do with the current API surface area.

During the process of tag helper / component discovery a compilation gets walked multiple times. We *could* introduce a cache during this process to avoid calling `AllInterfaces` multiple times per `INamedTypeSymbol`.